ma_search/data/shape.py

```diff
@@ -71,23 +93,128 @@ def polygon(self, tolerance=0.0, cachedOnly=True):
         Returns
         -------
         :obj:`shapely.Polygon`, :obj:`shapely.MultiPolygon` or None
+            Returns shape if it can be found, otherwise None.
+        """
+        if tolerance == 0.0:
+            return self.polygonFromGeoJson(self._path)
+        else:
+            suffix = f".{round(tolerance * 1e6)}.geojson"
+            path = (self._path.parent / self._path.stem).with_suffix(suffix)
+            exists = path.exists()
+
+            if exists:
+                return self.polygonFromGeoJson(path)
+            elif cachedOnly:
+                logger.error("Polygon does not exist for tolerance %s", tolerance)
+                return
+            elif not exists:
+                logger.info("Creating polygon %s with tolerance %s", self._uuid, tolerance)
+                full = self.polygonFromGeoJson(self._path)
+                simple = full.simplify(tolerance)
+                with path.open('w') as f:
+                    json.dump(mapping(simple), f)
+                return simple
+
+    def toGeoJson(self, **kwargs):
+        """Return geometry as as GeoJson
+        Parameters
+        ----------
+        kwargs : dict
+            kwargs to pass to Shape.polygon()
         """
-        return
+        geom = mapping(self.polygon(**kwargs))
+        geoJson = {"type": "Feature",
+                   "geometry": geom}
+        return geoJson
 
     ##
     #  Static Methods
     ##
 
     @staticmethod
-    def polygonFromGeoJson(geoJson):
-        """
+    def polygonFromGeoJson(data):
+        """Returns the polygon extracted from a geojson dict or file.
+
+        Parameters
+        ----------
+        data : dict or Path or str
+            Input data, either a single GeoJson feature, or only the geometry
+            section of a single feature. If str or Path: Path to the file
+            containing this dict.
+
+        Returns
+        -------
+        :obj:`shapely.Polygon`, :obj:`shapely.MultiPolygon` or None
+            Returns shape if it can be found, otherwise None.
         """
-        return
+        if isinstance(data, dict):
+            pass
+        elif isinstance(data, (Path, str)):
+            path = Path(data)
+            if not path.exists():
+                logger.error("%s does not exist", path)
+                return None
+            with path.open() as f:
+                data = json.load(f)
+        else:
+            logger.error("Input has to be dict, Path or string but is of type %s", type(data))
+            return None
+
+        if "geometry" in data:
+            data = data["geometry"]
+        try:
+            geom = shape(data)
+            if not isinstance(geom, (Polygon, MultiPolygon)):
+                raise TypeError("data is not a polygon but %s", type(geom))
```

Comment:
We should not raise any errors in this code as it will be running as a service, and this will crash the instance. All data validation should return error codes that we can handle and report back to the user.

Reply:
in try / except block, so it it caught already

Reply:
Yes, missed that. It's fine as it is.